Add missing events to hotplug

awels opened the pull request:
Signed-off-by: Alexander Wels <awels@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
Add missing events to VMI object when hotplugging volumes.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

davidvossel commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -1396,6 +1396,7 @@ func (c *VMIController) triggerHotplugPopulation(volume *virtv1.Volume, vmi *vir
 			return &syncErrorImpl{fmt.Errorf("Error creating hotplug population trigger pod %v", err), FailedCreatePodReason}
 		}
 	}
+	c.recorder.Eventf(vmi, k8sv1.EventTypeNormal, SuccessfulCreatePodReason, "Created hotplug trigger pod for volume %s", volume.Name)
```
shouldn't this go into the if statement right about this? that way we're sure we only get this event once per a pod creation?

awels replied:
Good point, fixed.

davidvossel reviewed (commented):
looks good, just one minor comment

davidvossel commented on pkg/virt-handler/vm.go:
```diff
@@ -551,22 +560,25 @@ func (d *VirtualMachineController) updateVMIStatus(vmi *v1.VirtualMachineInstanc
 								// mounted, and still in spec, and in phase we can change, update status to mounted.
 								volumeStatus.Phase = v1.HotplugVolumeMounted
 								volumeStatus.Message = fmt.Sprintf("Volume %s has been mounted in virt-launcher pod", volumeStatus.Name)
-								volumeStatus.Reason = "VolumeMountedToPod"
+								volumeStatus.Reason = VolumeMountedToPodReason
+								d.recorder.Event(vmi, k8sv1.EventTypeNormal, volumeStatus.Reason, volumeStatus.Message)
 							}
 						} else {
 							// Not mounted, check if the volume is in the spec, if not update status
 							if _, ok := specVolumeMap[volumeStatus.Name]; !ok && canUpdateToUnmounted(volumeStatus.Phase) {
 								// Not mounted.
 								volumeStatus.Phase = v1.HotplugVolumeUnMounted
 								volumeStatus.Message = fmt.Sprintf("Volume %s has been unmounted from virt-launcher pod", volumeStatus.Name)
-								volumeStatus.Reason = "VolumeUnMountedFromPod"
+								volumeStatus.Reason = VolumeUnMountedFromPodReason
+								d.recorder.Event(vmi, k8sv1.EventTypeNormal, volumeStatus.Reason, volumeStatus.Message)
 							}
 						}
 					} else {
 						// Successfully attached to VM.
 						volumeStatus.Phase = v1.VolumeReady
 						volumeStatus.Message = fmt.Sprintf("Successfully attach hotplugged volume %s to VM", volumeStatus.Name)
-						volumeStatus.Reason = "VolumeReady"
+						volumeStatus.Reason = VolumeReadyReason
+						d.recorder.Event(vmi, k8sv1.EventTypeNormal, volumeStatus.Reason, volumeStatus.Message)
```
I just realized that we're rebuilding the volume status here on each reconcile loop.  Similar to my other comment about ensuring we only fire an event once (when possible), we're going to want to do that here as well. These events should only fire if something changes, not if the same result is calculated again.

awels replied:
You make a very good point. Let me fix that.

davidvossel reviewed (commented):
sorry for the delayed review here. I had one more similar comment to my last one. The big take away here is under normal conditions we need to try and ensure we only fire events when something changes (if possible)

awels commented:
@davidvossel I updated the PR to not generate duplicate events, and added some unit tests to demonstrate this. While running manual tests I noticed we generate a lot of 'created attachment pod' events for the same attachment pod, but we also generate a ton of events for domain defined for instance, is that normal?

awels commented:
/test pull-kubevirt-e2e-k8s-1.17-rook-ceph

davidvossel commented on pkg/virt-handler/vm.go:
```diff
@@ -656,6 +687,93 @@ func (d *VirtualMachineController) migrationTargetUpdateVMIStatus(vmi *v1.Virtua
+		if newStatus.Phase != oldStatus.Phase {
+			d.recorder.Event(vmi, k8sv1.EventTypeNormal, newStatus.Reason, newStatus.Message)
+		}
+		delete(newStatusMap, newStatus.Name)
+	}
+	// Send events for any new statuses.
+	for _, v := range newStatusMap {
+		d.recorder.Event(vmi, k8sv1.EventTypeNormal, v.Reason, v.Message)
+	}
```
Thinking a head a bit. The newStatusMap is passed in by the caller function, and in this helper function items are deleted from it.

I don't know if there's any future use case where someone might want to use the newStatusMap, but i'd find it surprising if items were deleted by a helper function.

instead of deleting items, couldn't we create a new map in this helper which indicates what volume statuses have been handled, and then check that when iterating over the newStatusMap to determine what events are new?

awels replied:
Good point, modified the function to make a copy and delete from there.

davidvossel reviewed (approved):
/lgtm
/approve
